jbpm-task-assigning/kie-server-services-task-assigning/kie-server-services-task-assigning-planning/src/main/java/org/kie/server/services/taskassigning/planning/SolutionChangesBuilder.java

```diff
@@ -0,0 +1,348 @@
+import org.kie.server.services.taskassigning.core.model.solver.realtime.RemoveTaskProblemFactChange;
+import org.kie.server.services.taskassigning.core.model.solver.realtime.TaskPropertyChangeProblemFactChange;
+import org.kie.server.services.taskassigning.user.system.api.UserSystemService;
+import org.kie.server.api.model.taskassigning.TaskData;
+import org.optaplanner.core.impl.solver.ProblemFactChange;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static org.kie.server.services.taskassigning.core.model.Task.DUMMY_TASK_PLANNER_241;
+import static org.kie.server.services.taskassigning.core.model.Task.IS_NOT_DUMMY;
+import static org.kie.server.services.taskassigning.core.model.User.PLANNING_USER;
+import static org.kie.server.api.model.taskassigning.TaskStatus.Completed;
+import static org.kie.server.api.model.taskassigning.TaskStatus.Error;
+import static org.kie.server.api.model.taskassigning.TaskStatus.Exited;
+import static org.kie.server.api.model.taskassigning.TaskStatus.Failed;
+import static org.kie.server.api.model.taskassigning.TaskStatus.InProgress;
+import static org.kie.server.api.model.taskassigning.TaskStatus.Obsolete;
+import static org.kie.server.api.model.taskassigning.TaskStatus.Ready;
+import static org.kie.server.api.model.taskassigning.TaskStatus.Reserved;
+import static org.kie.server.api.model.taskassigning.TaskStatus.Suspended;
+import static org.kie.server.services.taskassigning.planning.SolutionBuilder.addInOrder;
+import static org.kie.server.services.taskassigning.planning.SolutionBuilder.fromTaskData;
+import static org.kie.server.services.taskassigning.planning.util.UserUtil.fromExternalUser;
+
+/**
+ * This class performs the calculation of the impact (i.e. the set of changes to be applied) on a solution given the
+ * updated information about the tasks in the jBPM runtime.
+ */
+public class SolutionChangesBuilder {
+
+    private static Logger LOGGER = LoggerFactory.getLogger(SolutionChangesBuilder.class);
+
+    private TaskAssigningSolution solution;
+
+    private List<TaskData> taskDataList;
+
+    private UserSystemService systemService;
+
+    private SolverHandlerContext context;
+
+    public SolutionChangesBuilder() {
+    }
+
+    public SolutionChangesBuilder withSolution(TaskAssigningSolution solution) {
+        this.solution = solution;
+        return this;
+    }
+
+    public SolutionChangesBuilder withTasks(List<TaskData> taskDataList) {
+        this.taskDataList = taskDataList;
+        return this;
+    }
+
+    public SolutionChangesBuilder withUserSystem(UserSystemService userSystemService) {
+        this.systemService = userSystemService;
+        return this;
+    }
+
+    public SolutionChangesBuilder withContext(SolverHandlerContext context) {
+        this.context = context;
+        return this;
+    }
+
+    public List<ProblemFactChange<TaskAssigningSolution>> build() {
```

Comment:
This method is *very* long. I think it would be good to split it into a couple sub-methods.